### PR TITLE
ConfigMaps for Spark Configuration of the Driver Container

### PR DESCRIPTION
**Pull request by tmckayus**
This specification outlines methods for exposing spark configuration
files through ConfigMaps in the container running the spark driver.

### CODE REVIEW

File: config-map-for-spark-driver.md [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+## Introduction
+
+This specification outlines methods for exposing spark configuration
+files through ConfigMaps in the container running the spark driver.
+
+## Problem statement
+
+Currently the only way to customize spark configuration files such
+as log4j.properties and spark-defaults.conf for spark driver
+applications is to create a custom oshinko s2i images which contain
```
Comment (tmckayus) [inline config-map-for-spark-driver.md]: typo, "a"

File: config-map-for-spark-driver.md [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+of a ConfigMap for custom spark driver configuration should be
+optional -- the default behavior should be to use the configuration
+defined in the image itself.
+
+Note that this is a similar but different problem from providing
+dynamic configuration of spark clusters (master and worker nodes).
+The OpenShift objects which define a spark cluster are created
+programmatically by oshinko, and so volumes and env vars can be
+setup up by oshinko during  cluster creation based on values set
+in JSON  payloads to oshinko-rest.
```
Comment (tmckayus) [inline config-map-for-spark-driver.md]: typo, double space

File: config-map-for-spark-driver.md [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+## Proposed solution
+
+ConfigMap objects mounted as volumes on the spark driver container
+would provide a way to modify the default configuration for the
+spark application.
+
+ConfigMap objects can store multiple named configuration files and
+be mounted to a particluar directory in a container. Consquently, it
+is possible to specify log4j.properties, spark-defaults.conf, and other
+files in a ConfigMap and mount them at a location such as /opt/spark/oshinko-conf
```
Comment (tmckayus) [inline config-map-for-spark-driver.md]: add single quotes on the path

File: config-map-for-spark-driver.md [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+be a parameter in the oshinko application templates.  It will have a
+default value of `oshinko-spark-driver-config` and will be a required
+value -- it must not be blank.
+
+There will be 2 ways for a user to provide a custom configuration for
+the driver:
+
+* use OpenShift facilities to edit the ConfigMap `oshinko-spark-driver-config`.
+  This will change the default spark driver configuration for any application
+  launched using the oshinko templates.
```
Comment (tmckayus) [inline config-map-for-spark-driver.md]: add "in the project", this is not site-wide

File: config-map-for-spark-driver.md [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+
+* create one or more ConfigMaps in the project to hold spark driver configurations.
+  Set the parameter in the oshinko templates to the name of one of these ConfigMaps
+  instead of `oshinko-spark-driver-config`.
```
Comment (tmckayus) [inline config-map-for-spark-driver.md]: when an application is launched

File: config-map-for-spark-driver.md [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+facilities before launching the application. If a specified ConfigMap
+does not exist when the application is launched, an error will be returned
+by OpenShift and the application will not run.
+
+### Alternatives
+
+There must be a method to dynamically specify the configuration
+of the spark driver.  
+
+One alternative mentiond here for completeness is to use clients
```
Comment (tmckayus) [inline config-map-for-spark-driver.md]: typo, mentiond

File: config-map-for-spark-driver.md [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+in the application pod to download a ConfigMap to the container
+and write the contents to an EmptyDir() volume in mounted in the
```
Comment (tmckayus) [inline config-map-for-spark-driver.md]: typo, extra "in"

File: config-map-for-spark-driver.md [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+be an additional parameter on the screen. Also, updates to the
+ConfigMap would not be written through to the pod automatically so
+behavior would not be consistent with ConfMaps used in spark cluster
+nodes.
+
+## Affected Components
+
+oshinko-s2i:
+
+* The application launcher has to modified to look in the
```
Comment (tmckayus) [inline config-map-for-spark-driver.md]: typo, missing "be"

**Review by elmiko — COMMENTED** [review]
aside from the nits which you have noted, i really like this. solid stuff Trevor.

just out of curiosity, will we add another feature later to allow users the option of adding config files to their s2i repos for auto-config magic?

**Comment by tmckayus** [comment]
yes, thesteve0 was working on a protoytype of that.   We could have 3 levels, and I think it makes sense to have all of them -- base config in the spark image, override from the git repo for your application during s2i if you like, then potentially override at launch time with a configmap if you like.  The outstanding question is how to treat these -- are they merges, or overwrites?  Do I copy my configs into the well-known config dir, overwriting the files I include but inheriting the other ones?  Or do I treat each config as a complete set and just move the SPARK_CONF_DIR env var around?

**Comment by elmiko** [comment]
good points @tmckayus 

we can leave those questions for the next iteration of this

**Pull request merged**